OpenTelemetry (S3M) Web Servers apps (#1795)
SumoLogic/sumologic-documentation: docs/integrations/web-servers/opentelemetry/haproxy-opentelemetry.md

By default, HAProxy logs are forwarded to Syslog. This needs to be changed to send the logs to files. Configuration in the file `/etc/haproxy/haproxy.cfg` is needed to be modified to send logs to files.

HAProxy logs have several levels of verbosity. To select a level, set your loglevel to one of the following:

- emerg - Errors such as running out of operating system file descriptors.
- alert - Some rare cases where something unexpected has happened, such as being unable to cache a response.
- info - TCP connection and http request details and errors.
- err - Errors such as being unable to parse a map file, being unable to parse the HAProxy configuration file, and when an operation on a stick table fails.
- warning - Certain important, but non-critical, errors such as failing to set a request header or failing to connect to a DNS nameserver.
- notice - Changes to a server's state, such as being UP or DOWN or when a server is disabled. Other events at startup, such as starting proxies and loading modules are also included. Health check logging, if enabled, also uses this level.

> - debug - Complete information, useful for development/testing.

Author: @himanshu219 @sumoanema @SumoLogic/sumoappdev This list doesn't match the one on [HAProxy's site](https://www.haproxy.com/blog/introduction-to-haproxy-logging/). Can you advise?

The Sumo Logic App for HAProxy helps you monitor activity in HAProxy. The preconfigured dashboards provide information about site visitors, including location of visitors, HTTP Error codes percentage, Backend and Frontend server statistics.

HAProxy logs are sent to Sumo Logic through OpenTelemetry [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Schematics.png')} alt="Schematics" />

## HAProxy Log Types

The app supports Logs from the open source version of HAProxy. The App is tested on the 2.3.9 version of HAProxy.

The HAProxy logs are generated in files as configured in the configuration file /etc/haproxy/haproxy.cfg ([learn more](https://www.haproxy.com/blog/introduction-to-haproxy-logging/)).

## Fields Create in Sumo Logic for HAProxy

Following are the [Fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of HAProxy App install if not already present.

> - **webengine.cluster.name** - User configured. Enter a name to identify the Haproxy cluster. This cluster name will be shown in the Sumo Logic dashboards.

Reviewer: ```suggestion
- **webengine.cluster.name** - User configured. Enter a name to identify the Haproxy cluster. This cluster name will be shown in the Sumo Logic dashboards.
- **`webengine.system`** - Has fixed value of **haproxy**
```

> - **sumo.datasource** - has fixed value of haproxy

Reviewer: ```suggestion
- **sumo.datasource** - Has fixed value of **haproxy**
```

## Prerequisite

This section provides instructions for configuring log collection for HAProxy running on a non-Kubernetes environment for the Sumo Logic App for HAProxy.

By default, HAProxy logs are forwarded to Syslog. This needs to be changed to send the logs to files. Configuration in the file /etc/haproxy/haproxy.cfg is needed to be modified to send logs to files.

> Haproxy logs have six levels of verbosity. To select a level, set loglevel to one of:

Author: ```suggestion
HAProxy logs have six levels of verbosity. To select a level, set loglevel to one of:
```

-   emerg - Errors such as running out of operating system file descriptors.
-   alert - Some rare cases where something unexpected has happened, such as being unable to cache a response.
-   info - TCP connection and http request details and errors.
-   err - Errors such as being unable to parse a map file, being unable to parse the HAProxy configuration file, and when an operation on a stick table fails.
-   warning - Certain important, but non-critical, errors such as failing to set a request header or failing to connect to a DNS nameserver.
-   notice - Changes to a server's state, such as being UP or DOWN or when a server is disabled. Other events at startup, such as starting proxies and loading modules are also included. Health check logging, if enabled, also uses this level.
-   debug - Complete information, useful for development/testing.

All logging settings are located in [Haproxy.conf](https://www.haproxy.com/blog/introduction-to-haproxy-logging/).

> For the dashboards to work properly, must set log format

Author: ```suggestion
For the dashboards to work properly, you'll need to set the log format.
```

```
%ci:%cp\ [%tr]\ %ft\ %b/%s\ %TR/%Tw/%Tc/%Tr/%Ta\ %ST\ %B\ %CC\ %CS\ %tsc\ %ac/%fc/%bc/%sc/%rc\ %sq/%bq\ %hr\ %hs\ %{+Q}r
```

1.  You can enable HAProxy logs to syslog by adding the following line in the global section of /etc/haproxy/haproxy.cfg file. This means that HAProxy will send its messages to rsyslog on 127.0.0.1.

```
global
   log 127.0.0.1  local2
```

> 1.  Now create a /`etc/rsyslog.d/haproxy.conf` file containing below lines.

Reviewer: this point is getting rendered as 1. It should be 2

```
local2.*    /var/log/haproxy.log
```

You can of course be more specific and create separate log files according to the level of messages:

```
local2.=info     /var/log/haproxy-info.log
local2.=notice     /var/log/haproxy-notice.log
```

> 1.  Restart HAProxy and rsyslog server to enforce configuration changes.

Reviewer: this point is getting rendered as 1. It should be 3

```
sudo service rsyslog restart
sudo service haproxy reload
```

## Collecting Logs, Metrics & installing HAProxy app

Here are the steps for Collecting Logs and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the "Use an existing Collector" option.
:::

> If you want to use an existing Otel Collector then this step can be skipped by selecting the option of using an existing Collector.

Reviewer: This line is repeating. We need to delete this.

If you want to create a new Collector please select **Add a new Collector** option.

Select the platform for which you want to install the Sumo OpenTelemetry Collector.

This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.  

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Collector.png')} alt="Collector" />

### Step 2: Configure integration

In this step we will be configuring the yaml required for HAProxy Collection.

Path of the log file configured to capture haproxy logs is needed to be given here.

The files are typically located in `/var/log/haproxy*.log`. If you're using a customized path, check the haproxy.conf file for this information.

> 

Reviewer: ```suggestion
You can add any custom fields which you want to tag along with the data ingested in sumo.
```

Click on the **Download YAML File** button to get the yaml file.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-YAML.png')} alt="YAML" />

### Step 3: Sending logs and metric to Sumo

Once you have the yaml file downloaded in step 2, follow the below steps based on the platform of the machine :

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},
    {label: 'macOS', value: 'macOS'},
  ]}>

<TabItem value="Linux">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Haproxy instance which needs to be monitored.
2.  restart the collector using
```sh
sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="macOS">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Haproxy instance which needs to be monitored.
2.  Restart the otelcol-sumo process using the below command 
```sh
otelcol-sumo --config /etc/otelcol-sumo/sumologic.yaml --config "glob:/etc/otelcol-sumo/conf.d/*.yaml"
```

</TabItem>
</Tabs>

After successful execution of the above command, Sumo will start receiving the data from your host machine.

> This will install the app to your Sumo Logic Org. The app consists of Dashboards.

Reviewer: ```suggestion
Press **Next**.This will install the app to your Sumo Logic Org. The app consists of Dashboards.
```

Panels will start to fill automatically. It's important to note that each panel slowly fills with data matching the time range query and received since the panel was created. Results won't immediately be available, but within 20 minutes, you'll see full graphs and map.

### Sample Log Messages

```
May 13 08:24:43 localhost haproxy[21813]:
27.2.81.92:64274 [13/May/2021:08:24:43.921] web-edupia.vn-4
```

### Sample Query

This query example is from the HAProxy - Overview dashboard > Top 5 URLs with Errors panel:

```

> %"proxy.cluster"=* %"sumo.datasource"=haproxy

Reviewer: ```suggestion
webengine.cluster.name=* %"sumo.datasource"=haproxy
```

---
id: haproxy-opentelemetry
title: HAProxy - OpenTelemetry Collector
sidebar_label: HAProxy - OTel Collector
description: Learn about the Sumo Logic OpenTelemetry App for HAProxy.
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<img src={useBaseUrl('img/integrations/web-servers/haproxy.png')} alt="Thumbnail icon" width="50"/> <img src={useBaseUrl('img/send-data/otel-color.svg')} alt="Thumbnail icon" width="45"/>

[HAProxy](https://docs.haproxy.org/2.6/intro.html) is open source software that provides a high availability load balancer and proxy server for TCP and HTTP-based applications that spreads requests across multiple servers.

The Sumo Logic App for HAProxy helps you monitor activity in HAProxy. The preconfigured dashboards provide information about site visitors, including location of visitors, HTTP Error codes percentage, Backend and Frontend server statistics.

HAProxy logs are sent to Sumo Logic through OpenTelemetry [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Schematics.png')} alt="Schematics" />

Reviewer 2: This image link is broken.

## HAProxy Log Types

The app supports Logs from the open source version of HAProxy. The App is tested on the 2.3.9 version of HAProxy.

The HAProxy logs are generated in files as configured in the configuration file /etc/haproxy/haproxy.cfg ([learn more](https://www.haproxy.com/blog/introduction-to-haproxy-logging/)).

## Fields Create in Sumo Logic for HAProxy

Following are the [Fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of HAProxy App install if not already present.

- **webengine.cluster.name** - User configured. Enter a name to identify the Haproxy cluster. This cluster name will be shown in the Sumo Logic dashboards.
- **`webengine.system`** - Has fixed value of **haproxy**
- **sumo.datasource** - Has fixed value of **haproxy**

> ## Prerequisite

Reviewer 2: ```suggestion
## Prerequisites
```

This section provides instructions for configuring log collection for HAProxy running on a non-Kubernetes environment for the Sumo Logic App for HAProxy.

> By default, HAProxy logs are forwarded to Syslog. This needs to be changed to send the logs to files. Configuration in the file /etc/haproxy/haproxy.cfg is needed to be modified to send logs to files.

Reviewer 2: ```suggestion
By default, HAProxy logs are forwarded to Syslog. This needs to be changed to send the logs to files. Configuration in the file `/etc/haproxy/haproxy.cfg` is needed to be modified to send logs to files.
```

-   debug - Complete information, useful for development/testing.

All logging settings are located in [Haproxy.conf](https://www.haproxy.com/blog/introduction-to-haproxy-logging/).
For the dashboards to work properly, you'll need to set the log format.

```
%ci:%cp\ [%tr]\ %ft\ %b/%s\ %TR/%Tw/%Tc/%Tr/%Ta\ %ST\ %B\ %CC\ %CS\ %tsc\ %ac/%fc/%bc/%sc/%rc\ %sq/%bq\ %hr\ %hs\ %{+Q}r
```

1.  You can enable HAProxy logs to syslog by adding the following line in the global section of /etc/haproxy/haproxy.cfg file. This means that HAProxy will send its messages to rsyslog on 127.0.0.1.

```
global
   log 127.0.0.1  local2
```

1.  Now create a /`etc/rsyslog.d/haproxy.conf` file containing below lines.

```
local2.*    /var/log/haproxy.log
```

You can of course be more specific and create separate log files according to the level of messages:

```
local2.=info     /var/log/haproxy-info.log
local2.=notice     /var/log/haproxy-notice.log
```

1.  Restart HAProxy and rsyslog server to enforce configuration changes.

```
sudo service rsyslog restart
sudo service haproxy reload
```

> ## Collecting Logs, Metrics & installing HAProxy app

Reviewer 2: ```suggestion
## Collecting Logs, Metrics and installing HAProxy app
```

Here are the steps for Collecting Logs and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the "Use an existing Collector" option.
:::

If you want to use an existing Otel Collector then this step can be skipped by selecting the option of using an existing Collector.

If you want to create a new Collector please select **Add a new Collector** option.

Select the platform for which you want to install the Sumo OpenTelemetry Collector.

This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.  

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Collector.png')} alt="Collector" />

Reviewer 2: This image link is broken.

### Step 2: Configure integration

In this step we will be configuring the yaml required for HAProxy Collection.

Path of the log file configured to capture haproxy logs is needed to be given here.

The files are typically located in `/var/log/haproxy*.log`. If you're using a customized path, check the haproxy.conf file for this information.
You can add any custom fields which you want to tag along with the data ingested in sumo.
Click on the **Download YAML File** button to get the yaml file.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-YAML.png')} alt="YAML" />

Reviewer 2: This image link is broken.

### Step 3: Sending logs and metric to Sumo

Once you have the yaml file downloaded in step 2, follow the below steps based on the platform of the machine :

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},

>     {label: 'macOS', value: 'macOS'},

Reviewer 2: ```suggestion
    {label: 'macOS', value: 'Mac'},
```

  ]}>

<TabItem value="Linux">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Haproxy instance which needs to be monitored.

> 2.  restart the collector using

Reviewer 2: ```suggestion
2.  Restart the collector using:
```

```sh
sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="macOS">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Haproxy instance which needs to be monitored.

> 2.  Restart the otelcol-sumo process using the below command 

Reviewer 2: ```suggestion
2.  Restart the otelcol-sumo process using the below command:
```

```sh
otelcol-sumo --config /etc/otelcol-sumo/sumologic.yaml --config "glob:/etc/otelcol-sumo/conf.d/*.yaml"
```

</TabItem>
</Tabs>

After successful execution of the above command, Sumo will start receiving the data from your host machine.

Press **Next**.This will install the app to your Sumo Logic Org. The app consists of Dashboards.

> Panels will start to fill automatically. It's important to note that each panel slowly fills with data matching the time range query and received since the panel was created. Results won't immediately be available, but within 20 minutes, you'll see full graphs and map.

Reviewer 2: ```suggestion
Panels will start to fill automatically. It's important to note that each panel slowly fills with data matching the time range query and received since the panel was created. Results won't immediately be available, but within 20 minutes, you'll see full graphs and maps.
```

### Threat Analysis

The **HAProxy - Threat Analysis** dashboard provides an at-a-glance view of threats to HAProxy servers on your network. Dashboard panels display the threat count over a selected time period, geographic locations where threats occurred, source breakdown, actors responsible for threats, severity, and a correlation of IP addresses, method, and status code of threats.

Use this dashboard to:

-   To gain insights and understand threats in incoming traffic and discover potential IOCs. Incoming traffic requests are analyzed using the [Sumo - Crowdstrikes](https://help.sumologic.com/docs/integrations/security-threat-detection/threat-intel-quick-analysis/#03_Threat-Intel-FAQ) threat feed.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Threat-Analysis.png')} alt="Threat Analysis" />

### Trends

The **HAProxy - Trends** dashboard provides an at-a-glance view of traffic to HAProxy servers on your network. Dashboard panels display the traffic count over one day time period, locations where traffic trends for visits by country one days time.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/HAProxy-OpenTelemetry/HAProxy-Trends.png')} alt="Trends" />

Reviewer 2: This image link is broken.
